README
The uefivars program is a Linux user-space utility to display the 
United Extensible Firmware Interface (UEFI) globally defined variables 
known to the kernel. You must be root to run this utility. 

efibootmgr is a Linux user-space application to modify the Intel
Extensible Firmware Interface (EFI) Boot Manager.  This application
can create and destroy boot entries, change the boot order, change
the next running boot option, and more.

The UEFI Specification is available from http://www.uefi.org.  See Section 
3.2 Table 8 of the UEFI specification for a full list of the globally 
defined variables.  Details on the EFI Boot Manager are available from the EFI
Specification, v1.02 or above, available from http://developer.intel.com.
 
Note: uefivars requires that the kernel was booted using (U)EFI
Note: efibootmgr requires that the kernel module efivars be loaded
prior to use.  'modprobe efivars' should do the trick.

Typical usage:
=============
 
[root @ localhost]# uefivars
Boot0000 (NV,BS,RT):* EFI USB Device   ACPI(a0841d0,0)PCI(1d,7)USB(2,0)
Boot0001 (NV,BS,RT):* EFI DVD/CDROM    ACPI(a0841d0,0)PCI(1f,2)ATAPI(0,0,0)
BootCurrent (BS,RT): 003D
BootOptionSupport (BS,RT): 0303
BootOrder (NV,BS,RT): 0000,0001
ConIn (NV,BS,RT):
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(30341d0,0)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7f010400)
/USBClass(ffff,ffff,3,1,1)
ConInDev (BS,RT):
/USBClass(ffff,ffff,3,1,1)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d3b419-2d9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4c7657e0-878e-9e2f-2808-83437f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4acf8bec-73a0-1dd0-e77e-2d887f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(30341d0,0)
ConOut (NV,BS,RT):
/ACPI(a0841d0,0)/PCI(0,1)/PCI(0,0)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7fff0400)
ConOutDev (BS,RT):
/ACPI(a0841d0,0)/PCI(0,1)/PCI(0,0)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d3b419-2d9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4c7657e0-878e-9e2f-2808-83437f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4acf8bec-73a0-1dd0-e77e-2d887fff0400)
ErrOut (NV,BS,RT):
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7fff0400)
ErrOutDev (BS,RT):
/ACPI(a0841d0,0)/PCI(0,1)/PCI(0,0)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d2f9be-0c9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(11d3b419-2d9a-9000-273f-c14d7f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4c7657e0-878e-9e2f-2808-83437f010400)
/ACPI(a0841d0,0)/PCI(0,1f)/ACPI(50141d0,1)/UART(115200,8,1,1)/Vendor(4acf8bec-73a0-1dd0-e77e-2d887fff0400)
Lang (NV,BS,RT): eng
LangCodes (BS,RT): eng,fra
PlatformLang (NV,BS,RT): en-US
PlatformLangCodes (BS,RT): en-US,fr-FR
Timeout (NV,BS,RT): 3 secs


== Using efibootmgr ==

usage: efibootmgr [options]
	-a | --active         sets bootnum active
	-A | --inactive       sets bootnum inactive
	-b | --bootnum XXXX   modify BootXXXX (hex)
	-B | --delete-bootnum delete bootnum (hex)
	-c | --create         create new variable bootnum and add to bootorder
	-d | --disk disk      (defaults to /dev/sda) containing loader
	-e | --edd [1|3|-1]   force EDD 1.0 or 3.0 creation variables, or guess
	-E | --device num     EDD 1.0 device number (defaults to 0x80)
	-g | --gpt            force disk w/ invalid PMBR to be treated as GPT 
	-H | --acpi_hid XXXX  set the ACPI HID (used with -i)
	-i | --iface name     create a netboot entry for the named interface
	-l | --loader name    (defaults to \elilo.efi)
	-L | --label label    Boot manager display label (defaults to "Linux")
	-n | --bootnext XXXX  set BootNext to XXXX (hex)
	-N | --delete-bootnext delete BootNext
	-o | --bootorder XXXX,YYYY,ZZZZ,...     explicitly set BootOrder (hex)
	-O | --delete-bootorder delete BootOrder
	-p | --part part       (defaults to 1) containing loader
	-q | --quiet           be quiet
	--test filename   don't write to NVRAM, write to filename
	-t | --timeout seconds   Boot manager timeout
	-T | --delete-timeout    delete Timeout value
	-u | --unicode | --UCS-2  pass extra args as UCS-2 (default is ASCII)
	-U | --acpi_uid XXXX    set the ACPI UID (used with -i)
	-v | --verbose         print additional information
	-V | --version         return version and exit
	-w | --write-signature  write unique sig to MBR if needed
        -@ | --append-binary-args   append extra variable args from
			   	    file (use - to read from stdin).  


Typical usage:
1) Root can use it to display the current Boot Manager settings.
  [root@localhost ~]# efibootmgr
  BootCurrent: 0004
  BootNext: 0003
  BootOrder: 0004,0000,0001,0002,0003
  Timeout: 30 seconds
  Boot0000* Diskette Drive(device:0)
  Boot0001* CD-ROM Drive(device:FF) 
  Boot0002* Hard Drive(Device:80)/HD(Part1,Sig00112233)   
  Boot0003* PXE Boot: MAC(00D0B7C15D91)               
  Boot0004* Linux

  This shows:
  BootCurrent - the boot entry used to start the currently running
  system.

  BootOrder - the boot order as would appear in the boot manager.  The
  boot manager tries to boot the first active entry on this list.  If
  unsuccessful, it tries the next entry, and so on.

  BootNext - the boot entry which is scheduled to be run on next boot.
  This superceeds BootOrder for one boot only, and is deleted by the
  boot manager after first use.  This allows you to change the next boot
  behavior without changing BootOrder.

  Timeout - the time in seconds between when the boot manager appears
  on the screen until when it automatically chooses the startup value
  from BootNext or BootOrder.

  Five boot entries (0000 - 0004), the active/inactive flag (* means
  active), and the name displayed on the screen.


2) An OS installer would call 'efibootmgr -c'.  This assumes that
   /boot/efi is your EFI System Partition, and is mounted at /dev/sda1.
   This creates a new boot option, called "Linux", and puts it at the top
   of the boot order list.  Options may be passed to modify the
   default behavior.  The default OS Loader is elilo.efi.

3) A system administrator wants to change the boot order.  She would
   call 'efibootmgr -o 3,4' to specify PXE boot first, then Linux
   boot.

4) A system administrator wants to change the boot order for the next
   boot only.  She would call 'efibootmgr -n 4' to specify that the
   Linux entry be taken on next boot.

5) A system administrator wants to delete the Linux boot option from
   the menu.  'efibootmgr -b 4 -B' deletes entry 4 and removes it
   from BootOrder.

6) A system administrator wants to create a boot option to network
   boot (PXE).  Unfortunately, this requires knowing a little more
   information about your system than can be easily found by
   efibootmgr, so you've got to pass additional information - the ACPI
   HID and UID values.  These can generally be found by using the EFI
   Boot Manager (in the EFI environment) to create a network boot
   entry, then using efibootmgr to print it verbosely.  Here's one example:

     Boot003* Acpi(PNP0A03,0)/PCI(5|0)/Mac(00D0B7F9F510) \
	ACPI(a0341d0,0)PCI(0,5)MAC(00d0b7f9f510,0)

   In this case, the ACPI HID is "0A0341d0" and the UID is "0".
   For the zx2000 gigE, the HID is "222F" and the UID is "500".
   For the rx2000 gigE, the HID is "0002" and the UID is "100".
   You create the boot entry with:
   'efibootmgr -c -i eth0 -H 222F -U 500 -L netboot'

Many other uses may be found.

Please direct any bugs, features, patches, etc. to Matt Domsch
<Matt_Domsch@dell.com>.







debian/changelog
efibootmgr (0.5.4-2ubuntu1) intrepid; urgency=low

  * Merge from debian unstable, remaining changes:
    - Build packages for lpia as well.
    - Set Ubuntu maintainer

 -- Bryce Harrington <bryce@ubuntu.com>  Mon, 09 Jun 2008 14:05:32 -0700

efibootmgr (0.5.4-2) unstable; urgency=low

  * build depend on libpci-dev instead of pcituils-dev, closes: #478374

 -- Bdale Garbee <bdale@gag.com>  Mon, 28 Apr 2008 18:12:40 -0700

efibootmgr (0.5.4-1) unstable; urgency=low

  * new upstream version

 -- Bdale Garbee <bdale@gag.com>  Sun, 24 Feb 2008 17:30:01 -0700

efibootmgr (0.5.3-3ubuntu1) hardy; urgency=low

  * Merge from debian unstable, remaining changes:
    - Build packages for lpia as well.
    - Set Ubuntu maintainer

 -- Bryce Harrington <bryce@ubuntu.com>  Wed, 12 Dec 2007 18:07:03 -0800

efibootmgr (0.5.3-3) unstable; urgency=low

  * fix FTBFS, closes: #450448
  * clean up a couple lintian warnings

 -- Bdale Garbee <bdale@gag.com>  Tue, 11 Dec 2007 23:04:09 -0700

efibootmgr (0.5.3-2ubuntu2) gutsy; urgency=low

  * Build packages for lpia as well.

 -- Adam Conrad <adconrad@ubuntu.com>  Wed, 15 Aug 2007 23:50:29 +1000

efibootmgr (0.5.3-2ubuntu1) feisty; urgency=low

  * Add -lz to src/efibootmgr/module.mk since pciutils now uses zlib.
    Fixes FTBFS.

 -- Tollef Fog Heen <tfheen@ubuntu.com>  Wed, 14 Mar 2007 12:08:06 +0100

efibootmgr (0.5.3-2build1) feisty; urgency=low

  * Rebuild for changes in the amd64 toolchain.

 -- Matthias Klose <doko@ubuntu.com>  Mon,  5 Mar 2007 01:15:22 +0000

efibootmgr (0.5.3-2) unstable; urgency=low

  * merge patch from Sam Hocevar (thanks!), closes: #389923

 -- Bdale Garbee <bdale@gag.com>  Fri, 20 Oct 2006 07:36:12 -0600

efibootmgr (0.5.3-1.1) unstable; urgency=medium

  * Non-maintainer upload.
  * src/include/gpt.h:
    + Fix GPT_HEADER_SIGNATURE declaration on 32bit architectures.
  * src/lib/efi.c:
    + Remove questionable "hack to allow include of ethtool.h".
    + Include asm/types.h so that proper types are defined (Closes: #389923).
  * debian/control:
    + Set policy to 3.7.2.
    + Uncapitalised short description.

 -- Sam Hocevar (Debian packages) <sam+deb@zoy.org>  Sun,  8 Oct 2006 14:21:38 +0200

efibootmgr (0.5.3-1) unstable; urgency=low

  * new upstream version, closes: #357884

 -- Bdale Garbee <bdale@gag.com>  Sat,  1 Apr 2006 22:00:02 -0800

efibootmgr (0.5.1-1) unstable; urgency=low

  * new upstream version. 

 -- Bdale Garbee <bdale@gag.com>  Thu, 10 Mar 2005 11:27:00 -0700

efibootmgr (0.5.0-1) unstable; urgency=low

  * new upstream version.  No code changes from 0.5.0-test4, which I called
    0.4.9 to avoid an epoch roll at this release... this upload just gets us
    back in version sync with the official release from upstream.

 -- Bdale Garbee <bdale@gag.com>  Sat, 28 Aug 2004 11:51:02 -0600

efibootmgr (0.4.9-0.sarge.2) testing; urgency=high

  * Base dependencies are frozen; this release reverts the portion of the
    test4 changes that added a dependency on pciutils.  Thanks to Dann Frazier
    for figuring this out and providing a suitable patch.  Closes: #268490

 -- Bdale Garbee <bdale@gag.com>  Fri, 27 Aug 2004 23:12:38 -0600

efibootmgr (0.4.9-0.sarge.1) testing; urgency=high

  * build 0.4.9-1 in testing chroot for sarge
  * new upstream version.  This is actually 0.5.0-test4, renumbered to avoid
    nastiness when 0.5.0 is released.  Upstream says 0.5.0 will be this code
    with a documentation change.  Closes: #258838, #235227
  * add pciutils-dev to build dependencies since we need libpci now
  * hack in a typedef to get a u64 definition for ethtool.h
  * urgency high because previous version fails with current 2.6 kernels

 -- Bdale Garbee <bdale@gag.com>  Wed, 18 Aug 2004 09:47:21 -0600

efibootmgr (0.4.9-1) unstable; urgency=high

  * new upstream version.  This is actually 0.5.0-test4, renumbered to avoid
    nastiness when 0.5.0 is released.  Upstream says 0.5.0 will be this code
    with a documentation change.  Closes: #258838, #235227
  * add pciutils-dev to build dependencies since we need libpci now
  * hack in a typedef to get a u64 definition for ethtool.h
  * urgency high because previous version fails with current 2.6 kernels

 -- Bdale Garbee <bdale@gag.com>  Wed, 18 Aug 2004 08:59:03 -0600

efibootmgr (0.4.2-4) unstable; urgency=low

  * add amd64 to the list of architectures to build on, closes: #249988

 -- Bdale Garbee <bdale@gag.com>  Sat, 22 May 2004 23:50:58 -0600

efibootmgr (0.4.2-3) unstable; urgency=low

  * patch from upstream to solve compilation problem, closes: #223146

 -- Bdale Garbee <bdale@gag.com>  Tue, 13 Jan 2004 11:20:20 +1030

efibootmgr (0.4.2-2) unstable; urgency=low

  * add i386 to list of supported architectures, since elilo depends on this
    package to support various options of the elilo shell script.  note that
    efibootmgr requires associated kernel support which probably isn't in 
    standard Debian kernels yet, but we'll deal with that when we get access
    to a real i386 system using EFI firmware...

 -- Bdale Garbee <bdale@gag.com>  Sat, 25 Oct 2003 15:23:59 -0600

efibootmgr (0.4.2-1) unstable; urgency=low

  * new upstream version

 -- Bdale Garbee <bdale@gag.com>  Thu,  4 Sep 2003 15:38:46 -0600

efibootmgr (0.4.1-1) unstable; urgency=low

  * new upstream version, closes: #165956
  *   this version can run non-root, closes: #163447
  *   this version checks if boot option was specified, closes: #163450
  * remove local copy of man page source and related processing now that
    man page is integrated with upstream release

 -- Bdale Garbee <bdale@gag.com>  Wed, 23 Oct 2002 16:05:53 -0600

efibootmgr (0.4.0-1) unstable; urgency=low

  * new upstream version, closes: #146544
  * updated man page from Dann Frazier <dannf@debian.org>, closes: #157213

 -- Bdale Garbee <bdale@gag.com>  Sun, 18 Aug 2002 20:40:48 -0600

efibootmgr (0.3.4-3) unstable; urgency=low

  * add man page from Dann Frazier <dannf@debian.org>, closes: #140514


src/include/efibootmgr.h
/*
  efibootmgr.h - Manipulates EFI variables as exported in /proc/efi/vars
 
  Copyright (C) 2001 Dell Computer Corporation <Matt_Domsch@dell.com>
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef _EFIBOOTMGR_H
#define _EFIBOOTMGR_H

typedef struct {
	int argc;
	char **argv;
	int optind;
	char *disk;
	char *iface;
	char *loader;
	char *label;
	char *bootorder;
	char *testfile;
	char *extra_opts_file;
	uint32_t part;
	int edd_version;
	int edd10_devicenum;
	int bootnum;
	int bootnext;
	int verbose;
	int active;
	uint32_t acpi_hid;
	uint32_t acpi_uid;
	unsigned int delete_boot:1;
	unsigned int delete_bootorder:1;
	unsigned int delete_bootnext:1;
	unsigned int quiet:1;
	unsigned int showversion:1;
	unsigned int create:1;
	unsigned int unicode:1;
	unsigned int write_signature:1;
	unsigned int forcegpt:1;
	unsigned int set_timeout:1;
	unsigned int delete_timeout:1;
	unsigned int zulu:1;
	unsigned short int timeout;
} efibootmgr_opt_t;


extern efibootmgr_opt_t opts;

#endif


src/uefivars/uefivars.c
/*
    uefivars.c - Lists UEFI variables as exported in /sys/firmware/efi

    Copyright (C) Finnbarr P. Murphy  <fpmATfpmurphy.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    Parts of this code come from efibootmgr.c which has the following 
    copyright:
 
    Copyright (C) 2001-2004 Dell, Inc. <Matt_Domsch@dell.com>
*/

#define _GNU_SOURCE

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <getopt.h>

#include "list.h"
#include "efi.h"
#include "efichar.h"
#include "unparse_path.h"
#include "efibootmgr.h"

/* Unused but needs to be defined for efi.c */
efibootmgr_opt_t opts;

/* allchar takes a unary function which operates on a char and a null
   terminated list of chars. If unary function is true for all char in
   char list, return true, else false. */
int allchar(int (*uchar)(int), char *p)
{
    for (;*p; p++) {
        if (uchar(*p) == 0)
            return 0;
    }
    
    return 1;
}

typedef struct _var_entry {
    struct dirent *name;
    uint16_t       num;
    efi_variable_t var_data;
    list_t         list;
} var_entry_t;

static LIST_HEAD(list_head);

void
free_vars(list_t *head)
{
    list_t *pos, *n;
    var_entry_t *boot;

    list_for_each_safe(pos, n, head) {
        boot = list_entry(pos, var_entry_t, list);
        list_del(&(boot->list));
        free(boot);
    }
}

void
read_vars(struct dirent **namelist,
          int num_boot_names,
          list_t *head)
{
    efi_status_t status;
    var_entry_t *entry;
    int i;

    if (!namelist) return;

    for (i=0; i < num_boot_names; i++) {
        if (namelist[i]) {
            entry = malloc(sizeof(var_entry_t));
            if (!entry) return;
            memset(entry, 0, sizeof(var_entry_t));

            status = read_variable(namelist[i]->d_name, &entry->var_data);
            if (status == EFI_SUCCESS) { 
                entry->name = namelist[i];
                list_add_tail(&entry->list, head);
            } else 
                free(entry);
        }
    }
    return;
}


void
free_dirents(struct dirent **ptr, 
             int num_dirents)
{
    int i;
    if (!ptr) return;
    for (i=0; i < num_dirents; i++) {
        if (ptr[i]) {
            free(ptr[i]);
            ptr[i] = NULL;
        }
    }
    free(ptr);
}


void
usage()
{
    printf("usage: uefivars [-v | --version]\n");
    printf("       uefivars [-u | --unknown]\n");
}


int
main(int argc, 
     char **argv)
{
    int c, first, unknownvar = 0;
    int option_index = 0;
    int num_all = 0;
    list_t *pos;
    var_entry_t *var;
    char desc[1032];
    char attributes[12];
    struct dirent  **all = NULL;

    static struct option long_options[] = {
        {"help", no_argument, 0, 'h'},
        {"unknown", no_argument, 0, 'u'},
        {"version", no_argument, 0, 'v'},
        {0, 0, 0, 0}
    };

    while ((c = getopt_long(argc, argv, "huv", long_options, &option_index)) != -1) {
        switch (c) {
            case 'h':
                 usage();
                 exit(EXIT_SUCCESS);
             case 'u':
                 unknownvar = 1;
                 break;
             case 'v':
                 fprintf(stdout, "version %s\n", EFIVARS_VERSION);
                 exit(EXIT_SUCCESS);
             default: /* '?' */
                 usage();
                 exit(EXIT_FAILURE);
        }
    }
        
    set_fs_kernel_calls();

    /* retrieve all entries */
    num_all = read_all_var_names(&all);
    read_vars(all, num_all, &list_head);

    /* loop through each entry and parse */
    list_for_each(pos, &list_head) {
        var = list_entry(pos, var_entry_t, list);
        efichar_to_char(desc, var->var_data.VariableName , sizeof(desc));

        /* parse attributes and build attributes text string */
        memset(attributes,0,12);
        first = 1;
        if (var->var_data.Attributes & EFI_VARIABLE_NON_VOLATILE) {
            if (first) {
               strcpy(attributes, "(");
               first = 0;
            } else strcat(attributes, ","); 
            strcat(attributes, "NV");
        }
        if (var->var_data.Attributes & EFI_VARIABLE_BOOTSERVICE_ACCESS) {
            if (first) {
                strcpy(attributes, "(");
                first = 0;
            } else strcat(attributes, ","); 
            strcat(attributes, "BS");
        }
        if (var->var_data.Attributes & EFI_VARIABLE_RUNTIME_ACCESS) {   
            if (first) {
                strcpy(attributes, "(");
                first = 0;
            } else strcat(attributes, ","); 
            strcat(attributes, "RT");
        }
        if (!first) strcat(attributes, ")");
        
        if (!strncmp(desc, "Boot", 4) && allchar(isxdigit, desc+4)) { 

            char description[80];
            EFI_LOAD_OPTION *load_option;
            EFI_DEVICE_PATH *path;
            char text_path[1024], *p;
            unsigned long optional_data_len=0;

            load_option = (EFI_LOAD_OPTION *) var->var_data.Data;
            efichar_to_char(description, load_option->description, sizeof(description));
            memset(text_path, 0, sizeof(text_path));
            path = load_option_path(load_option);

            if (var->name)
                printf("%.8s %s:", var->name->d_name, attributes);
            else
                printf("Boot%04X %s:", var->num, attributes);

            if (load_option->attributes & LOAD_OPTION_ACTIVE)
                printf("* ");
            else  
                printf("  ");

            printf("%s", description);

            unparse_path(text_path, path, load_option->file_path_list_length);
            optional_data_len = var->var_data.DataSize - load_option->file_path_list_length -
                                ((char *)path - (char *)load_option);
            if (optional_data_len) {
                 p = text_path;
                 p += strlen(text_path);
                 unparse_raw_text(p, ((uint8_t *)path) + load_option->file_path_list_length, optional_data_len);
            }
            printf("\t%s", text_path);
            printf("\n");
        } else if (!strcmp(desc, "BootCurrent")) {
            uint16_t *n = (uint16_t *)(var->var_data.Data);
            printf("BootCurrent %s: %04X\n", attributes, (unsigned int)*n);
        } else if (!strcmp(desc, "BootNext")) {
            uint16_t *n = (uint16_t *)(var->var_data.Data);
            printf("BootNext %s: %04X\n", attributes, (unsigned int)*n);
        } else if (!strcmp(desc, "BootOptionSupport")) {
            uint16_t *n = (uint16_t *)(var->var_data.Data);
            printf("BootOptionSupport %s: %04X\n", attributes, (unsigned int)*n);
        } else if (!strcmp(desc, "BootOrder")) {
            int i;
            uint16_t *data = (uint16_t *)&(var->var_data.Data);
            int *datasize = (int *)&(var->var_data.DataSize);
            int len = *datasize/sizeof(uint16_t);
            printf("BootOrder %s: ", attributes);
            for (i = 0; i < len; i++) {
                printf("%04X", data[i]);
                if (i < ( len - 1)) printf(",");
            }
            printf("\n");
        } else if (!strcmp(desc, "ConIn")) {
            printf("ConIn %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "ConInDev")) {
            printf("ConInDev %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "ConOut")) {
            printf("ConOut %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "ConOutDev")) {
            printf("ConOutDev %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "ErrOut")) {
            printf("ErrOut %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "ErrOutDev")) {
            printf("ErrOutDev %s:\n", attributes);
            efi_print_device_path(var->var_data.Data);
        } else if (!strcmp(desc, "Lang")) {
            int i;
            uint8_t *data = (uint8_t *)&(var->var_data.Data);
            printf("Lang %s: ", attributes);
            for (i = 0; i < 3; i++) {
                printf("%c", data[i]);
            }
            printf("\n");
        } else if (!strcmp(desc, "LangCodes")) {
            int i;
            uint8_t *data = (uint8_t *)&(var->var_data.Data);
            printf("LangCodes %s: ", attributes);
            while (1) {
               for (i = 0; i < 3; i++) {
                   printf("%c", *data++);
               }
               if (*data < 32) break;
               printf(",");
            }
            printf("\n");
        } else if (!strcmp(desc, "PlatformLang")) {
            int i;
            uint8_t *data = (uint8_t *)&(var->var_data.Data);
            printf("PlatformLang %s: ", attributes);
            for (i = 0; i < 5; i++) {
                printf("%c", data[i]);
            }
            printf("\n");
        } else if (!strcmp(desc, "PlatformLangCodes")) {
            int i;
            uint8_t *data = (uint8_t *)&(var->var_data.Data);
            printf("PlatformLangCodes %s: ", attributes);
            while (1) {
               for (i = 0; i < 5; i++) {
                   printf("%c", *data++);
               }
               if (*data < 32) break;
               if (*data == ';') {
                   printf(",");
                   data++;
               }
            }
            printf("\n");
        } else if (!strcmp(desc, "Timeout")) {
            uint16_t *n = (uint16_t *)(var->var_data.Data);
            printf("Timeout %s: %d secs\n", attributes, (int)*n);
        } else if (unknownvar) {
            printf("%s %s: UNKNOWN\n", desc, attributes);
        }
    }

    free_dirents(all, num_all);
    free_vars(&list_head);

    return 0;
}

